🐞 - Invalid migration on tui-editor repo

- [x] TuiDropdownService


I have 

```ts
import {TuiDropdownService} from '@taiga-ui/core';
```

```
✘ [ERROR] TS2724: '"@taiga-ui/core"' has no exported member named 'TuiDropdownService'. Did you mean 'TuiDropdownDriver'? [plugin angular-compiler]

    projects/editor/components/editor/portal/editor-portal.directive.ts:2:23:
      2 │ import {tuiAsViewport, TuiDropdownService, TuiRectAccessor} from '@...
        ╵                        ~~~~~~~~~~~~~~~~~~
```

It should be 

```ts
import {TuiPopupService} from '@taiga-ui/core';
```

- [x] The legacy package does not have addon-mobile in peer deps


```

Application bundle generation complete. [0.625 seconds]

✘ [ERROR] Could not resolve "@taiga-ui/addon-mobile/directives/ripple"
    node_modules/@taiga-ui/legacy/fesm2022/taiga-ui-legacy-components-mobile-dialog.mjs:4:26:
      4 │ import { TuiRipple } from '@taiga-ui/addon-mobile/directives/ripple';
        ╵                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  You can mark the path "@taiga-ui/addon-mobile/directives/ripple" as external to exclude it from the bundle, which will remove this error and leave the unresolved path in the bundle.
/Users/m.ivanov4/WebstormProjects/editor/node_modules/esbuild/lib/main.js:1463
  let error = new Error(text);
              ^
Error: Build failed with 1 error:
node_modules/@taiga-ui/legacy/fesm2022/taiga-ui-legacy-components-mobile-dialog.mjs:4:26: ERROR: Could not resolve "@taiga-ui/addon-mobile/directives/ripple"
    at failureErrorWithLog (/Users/m.ivanov4/WebstormProjects/editor/node_modules/esbuild/lib/main.js:1463:15)
    at /Users/m.ivanov4/WebstormProjects/editor/node_modules/esbuild/lib/main.js:924:25
    at /Users/m.ivanov4/WebstormProjects/editor/node_modules/esbuild/lib/main.js:1341:9
    at process.processTicksAndRejections (node:internal/process/task_queues:103:5) {
  errors: [Getter/Setter],
  warnings: [Getter/Setter]
}
Node.js v24.13.0

```

- [x] Invalid migration for tui-textarea


I have

```html
<tui-textarea
                        [expandable]="true"
                        [ngModel]="control.value"
                        [ngModelOptions]="{updateOn: 'blur'}"
                        (ngModelChange)="control.setValue($event)"
                    />
```

But 

```

✘ [ERROR] NG8002: Can't bind to 'expandable' since it isn't a known property of 'tui-textarea'.
1. If 'tui-textarea' is an Angular component and it has 'expandable' input, then verify that it is included in the '@Component.imports' of this component.
2. If 'tui-textarea' is a Web Component then add 'CUSTOM_ELEMENTS_SCHEMA' to the '@Component.schemas' of this component to suppress this message.
3. To allow any property add 'NO_ERRORS_SCHEMA' to the '@Component.schemas' of this component. [plugin angular-compiler]

    projects/demo/src/app/pages/starter/index.html:52:24:
      52 │                         [expandable]="true"
         ╵                         ~~~~~~~~~~~~~~~~~~~

  Error occurs in the template of component TuiEditorStarter.
```

- [x] Invalid migration for tuiGetClosestFocusable, tuiIsNativeMouseFocusable


Before I have

```ts

    private findNextTool(wrapper?: HTMLElement | null): HTMLElement | null {
        return !wrapper || tuiIsNativeMouseFocusable(wrapper)
            ? (wrapper ?? null)
            : tuiGetClosestFocusable({
                  initial: wrapper,
                  root: this.el,
                  keyboard: false,
              });
    }
```


```ts
import {
    tuiIsNativeMouseFocusable,
} from '@taiga-ui/cdk';
```


But 

```

✘ [ERROR] TS2353: Object literal may only specify known properties, and 'keyboard' does not exist in type 'TuiGetClosestFocusableOptions'. [plugin angular-compiler]

    projects/editor/components/toolbar-host/toolbar-navigation-manager.directive.ts:89:18:
      89 │                   keyboard: false,
         ╵                   ~~~~~~~~
```

```

✘ [ERROR] TS2305: Module '"@taiga-ui/cdk"' has no exported member 'tuiIsNativeMouseFocusable'. [plugin angular-compiler]

    projects/editor/components/toolbar-host/toolbar-navigation-manager.directive.ts:6:4:
      6 │     tuiIsNativeMouseFocusable,
        ╵     ~~~~~~~~~~~~~~~~~~~~~~~~~
```

- [x] Invalid migration for tuiRepeatTimes


```ts
import {
    ChangeDetectionStrategy,
    Component,
    effect,
    forwardRef,
    inject,
    TemplateRef,
    viewChild,
} from '@angular/core';
import {WA_WINDOW} from '@ng-web-apis/common';
import {TuiRepeatTimes} from '@taiga-ui/cdk';
import { tuiDropdown, TuiDropdownDirective, tuiDropdownOpen, tuiGetViewportWidth, TuiTextfieldDropdownDirective, TuiWithDropdownOpen, TuiInput } from '@taiga-ui/core';
import {type TuiEditorOptions} from '@taiga-ui/editor/common';
import {type TuiLanguageEditor} from '@taiga-ui/i18n';

import {TuiToolbarTool} from '../tool';
import {TuiToolbarButtonTool} from '../tool-button';

const MAX_COLS_NUMBER = 15;
const MAX_ROWS_NUMBER = 15;
const MIN_DISTANCE_PX = 70;

@Component({
    selector: 'button[tuiInsertTableTool]',
    imports: [TuiRepeatTimes, TuiInput],
    template: `
        {{ tuiHint() }}

        <ng-container *tuiDropdown>
            <div class="t-size-selector">
                <div
                    *tuiRepeatTimes="let x of columnsNumber"
                    class="t-column"
                >
                    <div
                        *tuiRepeatTimes="let y of rowsNumber"
                        class="t-cell"
                        [class.t-cell_hovered]="tableSelectHovered(y, x)"
                        (click)="addTable(tableSize)"
                        (mouseenter)="updateCurrentSize(y + 1, x + 1, $event)"
                    ></div>
                </div>
                <div class="t-description">
                    {{ tableSize.cols }}&#215;{{ tableSize.rows }}
                </div>
            </div>
        </ng-container>
    `,
    styles: `
        .t-size-selector {
            display: block;
            padding: 0.75rem;
        }

        .t-cell {
            display: inline-block;
            background-color: var(--tui-background-base);
            inline-size: 1.25rem;
            block-size: 1.25rem;
            border-radius: 0.25rem;
            margin: 0.125rem;
            border: 0.0625rem solid var(--tui-border-normal);
            cursor: pointer;
        }

        .t-cell_hovered {
            background-color: var(--tui-background-base-alt);
        }

        .t-column {
            white-space: nowrap;
        }

        .t-description {
            text-align: center;
        }
    `,
    changeDetection: ChangeDetectionStrategy.OnPush,
    hostDirectives: [TuiToolbarButtonTool, TuiDropdownDirective, TuiWithDropdownOpen],
    host: {'[attr.automation-id]': '"toolbar__insert-table-button"'},
})
export class TuiInsertTableButtonTool extends TuiToolbarTool {
    private readonly win = inject(WA_WINDOW);
    protected readonly dropdown = tuiDropdown(null);
    protected readonly open = tuiDropdownOpen();

    protected tableSize = {
        rows: 1,
        cols: 1,
    };

    protected readonly template = viewChild(
        forwardRef(() => TuiTextfieldDropdownDirective),
        {read: TemplateRef},
    );

    protected readonly templateEffect = effect(() => {
        this.dropdown.set(this.template());
    });

    protected get columnsNumber(): number {
        return Math.min(Math.max(3, this.tableSize.cols + 1), MAX_COLS_NUMBER);
    }

    protected get rowsNumber(): number {
        return Math.min(Math.max(3, this.tableSize.rows + 1), MAX_ROWS_NUMBER);
    }

    protected getIcon(icons: TuiEditorOptions['icons']): string {
        return icons.insertTable;
    }

    protected getHint(texts?: TuiLanguageEditor['toolbarTools']): string {
        return this.open() ? '' : (texts?.insertTable ?? '');
    }

    protected addTable({rows, cols}: {cols: number; rows: number}): void {
        this.editor()?.enter(); // @note: clear previous styles

        const prevLine = this.editor()?.state?.selection.anchor;

        // @note: don't use `setHardBreak`,
        // it inherits styles of previous lines
        // required two line after
        this.editor()?.enter();
        this.editor()?.enter();

        this.editor()?.setTextSelection(prevLine ?? 0);
        this.editor()?.insertTable(rows, cols);
    }

    protected tableSelectHovered(x: number, y: number): boolean {
        return x < this.tableSize.rows && y < this.tableSize.cols;
    }

    protected updateCurrentSize(rows: number, cols: number, event: MouseEvent): void {
        if (tuiGetViewportWidth(this.win) - event.clientX > MIN_DISTANCE_PX) {
            this.tableSize = {rows, cols};
            this.cd.detectChanges();
        }
    }
}
```

```

✘ [ERROR] TS-991010: 'imports' must be an array of components, directives, pipes, or NgModules.
  Value could not be determined statically. [plugin angular-compiler]

    projects/editor/tools/buttons/insert-table.ts:25:14:
      25 │     imports: [TuiRepeatTimes, TuiInput],
         ╵               ~~~~~~~~~~~~~~

  Unknown reference.

    projects/editor/tools/buttons/insert-table.ts:25:14:
      25 │     imports: [TuiRepeatTimes, TuiInput],
         ╵               ~~~~~~~~~~~~~~
```

- [x] Invalid migration for TuiColorSelectorModule


```
import {TuiColorSelectorModule} from '@taiga-ui/legacy';
```

```html
    <tui-color-selector
        [color]="selectedColor"
        [colors]="colors"
        (colorChange)="onValueChange($event)"
    />
```

- [x]  Invalid migration `@ng-web-apis/common`

```ts
import {LOCAL_STORAGE} from '@ng-web-apis/common';
import {WINDOW} from '@ng-web-apis/common';
```

- [x]  Invalid migration for tui-textfield


Before

```ts
<tui-textarea
    class="tui-space_top-5"
    [ngModel]="markdown"
    [style.min-height.rem]="30"
    (ngModelChange)="markdown$.next($event)"
>
    Markdown
</tui-textarea>
```

After

```
<tui-textfield class="tui-space_top-5" [style.min-height.rem]="30" (ngModelChange)="markdown$.next($event)">
<label tuiLabel>Markdown</label>
<textarea tuiTextarea [ngModel]="markdown"></textarea>
</tui-textfield>
```

Expected behaviour

```
<tui-textfield class="tui-space_top-5" [style.min-height.rem]="30">
<label tuiLabel>Markdown</label>
<textarea tuiTextarea [ngModel]="markdown" (ngModelChange)="markdown$.next($event)"></textarea>
</tui-textfield>
```

- [x] Invalid migration tui-textarea



Before

```ts
 <tui-textarea
    [expandable]="true"
    [ngModel]="control.value"
    [ngModelOptions]="{updateOn: 'change'}"
    (ngModelChange)="control.setValue($event)"
/>
```

After

Nothing changes, but

```

✘ [ERROR] NG8002: Can't bind to 'expandable' since it isn't a known property of 'tui-textarea'.
1. If 'tui-textarea' is an Angular component and it has 'expandable' input, then verify that it is included in the '@Component.imports' of this component.
2. If 'tui-textarea' is a Web Component then add 'CUSTOM_ELEMENTS_SCHEMA' to the '@Component.schemas' of this component to suppress this message.
3. To allow any property add 'NO_ERRORS_SCHEMA' to the '@Component.schemas' of this component. [plugin angular-compiler]

    projects/demo/src/app/pages/starter/index.html:52:24:
      52 │                         [expandable]="true"
         ╵                         ~~~~~~~~~~~~~~~~~~~
```

- [x] Invalid migration for tuiDocAPIItem



Before

```html
<ng-template
    documentationPropertyMode="input"
    documentationPropertyName="ngModel"
    documentationPropertyType="string"
    [documentationPropertyValue]="control.value"
    (documentationPropertyValueChange)="control.setValue($event)"
>
    Example pass HTML code
</ng-template>
```

After

```html
<tr
    name="[ngModel]"
    tuiDocAPIItem
    type="string"
    [value]="control.value"
    (valueChange)="control.setValue($event)"
>
    Example pass HTML code
</tr>
```

```

✘ [ERROR] NG5: Argument of type 'string | null | undefined' is not assignable to parameter of type 'string | null'.
  Type 'undefined' is not assignable to type 'string | null'. [plugin angular-compiler]

    projects/demo/src/app/pages/starter/index.html:110:44:
      110 │             (valueChange)="control.setValue($event)"
```


- [x] Invalid migration for tui-accordion-item


Before

```html
<tui-accordion [rounded]="false">
    <tui-accordion-item [open]="true">
        Output

        <ng-template tuiAccordionItemContent>
            <pre><code>{{ group.value|json }}</code></pre>
        </ng-template>
    </tui-accordion-item>
</tui-accordion>
```

After

```html
<tui-accordion [rounded]="false">
    <button [tuiAccordion]="true">Output</button>
    <tui-expand><pre><code>{{ group.value|json }}</code></pre></tui-expand>
</tui-accordion>
```

```
✘ [ERROR] NG8002: Can't bind to 'rounded' since it isn't a known property of 'tui-accordion'.
1. If 'tui-accordion' is an Angular component and it has 'rounded' input, then verify that it is included in the '@Component.imports' of this component.
2. If 'tui-accordion' is a Web Component then add 'CUSTOM_ELEMENTS_SCHEMA' to the '@Component.schemas' of this component to suppress this message.
3. To allow any property add 'NO_ERRORS_SCHEMA' to the '@Component.schemas' of this component. [plugin angular-compiler]

    projects/demo/src/app/pages/toolbar/shared/examples/1/index.html:40:15:
      40 │ <tui-accordion [rounded]="false">
         ╵                ~~~~~~~~~~~~~~~~~
```

- [x] Invalid migration for tuiDropdownOpen


Before

```ts
import {
    tuiDropdown
} from '@taiga-ui/core';

export class Test {
    protected readonly open = tuiDropdownOpen();
}
```

```

✘ [ERROR] TS2724: '"@taiga-ui/core"' has no exported member named 'tuiDropdownOpen'. Did you mean 'TuiDropdownOpen'? [plugin angular-compiler]

    projects/editor/tools/buttons/size.ts:12:57:
      12 │ ..., TuiDropdownDirective, tuiDropdownOpen, TuiOption, TuiTextfiel...
         ╵                            ~~~~~~~~~~~~~~~

  'TuiDropdownOpen' is declared here.
```

Expected

```
    protected readonly open = inject(TuiDropdownOpen).open
```

- [x] Invalid migration tuiHintOptionsProvider


Before

```ts
    providers: [tuiHintOptionsProvider({direction: ['top-left', 'top', 'right']})],
```

```

✘ [ERROR] TS2322: Type '"top-left"' is not assignable to type 'TuiHintDirection'. [plugin angular-compiler]

    projects/editor/tools/tool-button.ts:13:44:
      13 │ ...tuiHintOptionsProvider({direction: ['top-left', 'top', 'right']}),
         ╵                                        ~~~~~~~~~~


✘ [ERROR] TS2322: Type '"right"' is not assignable to type 'TuiHintDirection'. [plugin angular-compiler]

    projects/editor/tools/tool-button.ts:13:63:
      13 │ ...tuiHintOptionsProvider({direction: ['top-left', 'top', 'right']}),
```

- [x] Invalid migration for TuiTextfieldDropdownDirective


Before

```ts
import {
    TuiTextfieldDropdownDirective,
} from '@taiga-ui/core';

///...


@ViewChild(forwardRef(() => TuiTextfieldDropdownDirective), {read: TemplateRef})
protected set template(template: PolymorpheusContent) {
  this.dropdown.set(template);
}

```


After

```
✘ [ERROR] TS2724: '"@taiga-ui/core"' has no exported member named 'TuiTextfieldDropdownDirective'. Did you mean 'TuiTextfieldOptionsDirective'? [plugin angular-compiler]

    projects/editor/components/editor/editor.component.ts:47:4:
      47 │     TuiTextfieldDropdownDirective,
         ╵     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

- [x] Invalid migration for TUI_RANGE


```ts
import {
    TUI_RANGE
} from '@taiga-ui/cdk';
```

```


✘ [ERROR] TS2724: '"@taiga-ui/cdk"' has no exported member named 'TUI_RANGE'. Did you mean 'tuiInRange'? [plugin angular-compiler]

    projects/editor/components/editor/dropdown/dropdown-toolbar.directive.ts:14:4:
      14 │     TUI_RANGE,
         ╵     ~~~~~~~~~
```

Should migrate to 

```ts
    protected range = isPlatformBrowser(inject(PLATFORM_ID))
        ? new Range()
        : ({} as unknown as Range);
```

- [x]  Invalid migration for TuiHintDirective


Before

```ts
 protected readonly tuiHint = tuiDirectiveBinding(
        TuiHintDirective,
        'tuiHint',
        computed((texts = this.texts()) => this.getHint(texts)),
    );

    protected readonly tuiHintManual = tuiDirectiveBinding(
        TuiHintManual,
        'tuiHintManual',
        !this.isMobile && null,
    );
```

Should be 

```ts
    protected readonly tuiHint = tuiDirectiveBinding(
        TuiHintDirective,
        'content',
        computed((texts = this.texts()) => this.getHint(texts)),
    );

   protected readonly tuiHintManual = tuiDirectiveBinding(
        TuiHintManual,
        'visible',
        !this.isMobile && null,
    );
```

--- a/projects/cdk/schematics/ng-update/utils/replace-function-parameters.ts
+++ b/projects/cdk/schematics/ng-update/utils/replace-function-parameters.ts
@@ -39,17 +39,33 @@ function replaceFunctionParameter(item: ReplacementFunctionParameter): void {
             }
 
             if (item.valueReplacer) {
-                const replacement = property.getLastChildIfKind(SyntaxKind.StringLiteral);
+                const stringLiteral = property.getLastChildIfKind(
+                    SyntaxKind.StringLiteral,
+                );
 
-                if (!replacement) {
-                    return;
+                if (stringLiteral) {
+                    item.valueReplacer.forEach(({from, to}) => {
+                        if (stringLiteral.getLiteralValue() === from) {
+                            stringLiteral.setLiteralValue(to);
+                        }
+                    });
+                } else {
+                    const arrayLiteral = property.getLastChildIfKind(
+                        SyntaxKind.ArrayLiteralExpression,
+                    );
+
+                    arrayLiteral?.getElements().forEach((element) => {
+                        if (!Node.isStringLiteral(element)) {
+                            return;
+                        }
+
+                        item.valueReplacer!.forEach(({from, to}) => {
+                            if (element.getLiteralValue() === from) {
+                                element.setLiteralValue(to);
+                            }
+                        });
+                    });
                 }
-
-                item.valueReplacer.forEach(({from, to}) => {
-                    if (replacement.getLiteralValue() === from) {
-                        replacement.setLiteralValue(to);
-                    }
-                });
             }
 
             if (!renameTo) {

--- a/projects/cdk/schematics/ng-update/v5/tests/schematic-migrate-attr-values.spec.ts
+++ b/projects/cdk/schematics/ng-update/v5/tests/schematic-migrate-attr-values.spec.ts
@@ -372,6 +372,26 @@ describe('ng-update start/end instead of left/right', () => {
     );
 
     it(
+        'replaces for direction array on tuiHintOptionsProvider',
+        migrate({
+            component: /* TypeScript */ `
+                import {tuiHintOptionsProvider} from '@taiga-ui/core';
+
+                tuiHintOptionsProvider({direction: ['top-left', 'top', 'right']});
+                tuiHintOptionsProvider({
+                    direction: [
+                        'bottom-left',
+                        'bottom',
+                        'bottom-right',
+                        'left',
+                        'right-top',
+                    ],
+                });
+            `,
+        }),
+    );
+
+    it(
         'replaces for direction on tuiHintOptionsProvider',
         migrate({
             component: /* TypeScript */ `

